src/cache.rs

```diff
@@ -0,0 +1,341 @@
+/// Core logic for cache files. Used by `crate::actors::cache`.
+///
+/// Terminology:
+/// - positive cache item: A cache item that represents the presence of something. E.g. we
+///   succeeded in downloading an object file and cached that file.
+/// - Negative cache item: A cache item that represents the absence of something. E.g. we encountered a 404 while trying to download a file, and cached that fact. Represented by an empty file.
+///
+/// - item age: When the item was computed and created.
+/// - item last used: Last time this item was involved in a cache hit.
+///
+/// TODO:
+/// * We want to try upgrading derived caches without pruning them. This will likely require the concept of a content checksum (which would just be the cache key of the object file that would be used to create the derived cache.
+use std::fs::{create_dir_all, read_dir, remove_file, File, OpenOptions};
+use std::io::{self, Read};
+use std::path::{Path, PathBuf};
+use std::time::{Duration, SystemTime};
+
+use failure::Fail;
+
+use sentry::integrations::failure::capture_fail;
+use symbolic::common::ByteView;
+
+use crate::config::CacheConfig;
+use crate::logging::LogError;
+use crate::types::Scope;
+
+/// Content of symcaches/cficaches whose writing failed.
+///
+/// Items with this value will be considered expired after the next process restart, or will be
+/// pruned once `symbolicator cleanup` runs. Independently of any `max_age` or `max_last_used`.
+///
+/// The malformed state is useful for failed computations that are unlikely to succeed before the
+/// next deploy. For example, symcache writing may fail due to an object file symbolic can't parse
+/// yet.
+pub const MALFORMED_MARKER: &[u8] = b"malformed";
+
+#[derive(Debug, Fail, derive_more::From)]
+pub enum CleanupError {
+    #[fail(display = "No caching configured! Did you provide a path to your config file?")]
+    NoCachingConfigured,
+
+    #[fail(display = "Not a file")]
+    NotAFile,
+
+    #[fail(display = "Failed to access filesystem")]
+    Io(#[fail(cause)] io::Error),
+}
+
+/// Utilities for a sym/cfi or object cache.
+#[derive(Debug, Clone)]
+pub struct Cache {
+    /// Cache identifier used for metric names.
+    pub name: &'static str,
+
+    /// Directory to use for storing cache items. Will be created if it does not exist.
+    pub cache_dir: Option<PathBuf>,
+
+    /// Time when this process started.
+    start_time: SystemTime,
```

Comment:
Maybe make this public too?

Reply:
Ideally users wouldn't need to care about this. Tests already have access and for those I can imagine it being useful.

Reply:
Fixed by making everything private.